probe/plugins/registry.go

```diff
@@ -180,23 +212,140 @@ func (r *Registry) Name() string { return "plugins" }
 func (r *Registry) Report() (report.Report, error) {
 	rpt := report.MakeReport()
 	// All plugins are assumed to (and must) implement reporter
-	r.ForEach(func(plugin *Plugin) {
+	r.forEach(&r.lock, func(plugin *Plugin) {
 		pluginReport, err := plugin.Report()
 		if err != nil {
 			log.Errorf("plugins: %s: /report error: %v", plugin.socket, err)
 		}
+		if plugin.Implements("controller") {
+			r.updateAndRegisterControlsInReport(&pluginReport)
+		}
 		rpt = rpt.Merge(pluginReport)
 	})
 	return rpt, nil
 }
 
+func (r *Registry) updateAndRegisterControlsInReport(rpt *report.Report) {
+	key := rpt.Plugins.Keys()[0]
+	spec, _ := rpt.Plugins.Lookup(key)
+	pluginID := spec.ID
+	topologies := topologyPointers(rpt)
+	r.handlerRegistry.Lock()
+	defer r.handlerRegistry.Unlock()
+	r.rmPluginHandlers(pluginID)
+	for _, topology := range topologies {
+		r.updateAndRegisterControlsInTopology(pluginID, topology)
+	}
+}
+
+func topologyPointers(rpt *report.Report) []*report.Topology {
+	// We cannot use rpt.Topologies(), because it makes a slice of
+	// topology copies and we need original locations to modify
+	// them.
+	return []*report.Topology{
+		&rpt.Endpoint,
+		&rpt.Process,
+		&rpt.Container,
+		&rpt.ContainerImage,
+		&rpt.Pod,
+		&rpt.Service,
+		&rpt.Deployment,
+		&rpt.ReplicaSet,
+		&rpt.Host,
+		&rpt.Overlay,
+	}
+}
+
+func (r *Registry) rmPluginHandlers(pluginID string) {
+	for _, controlID := range r.controlsByPlugin[pluginID] {
+		fakeID := fakeControlID(pluginID, controlID)
+		r.handlerRegistry.RmLocked(fakeID)
+	}
+	delete(r.controlsByPlugin, pluginID)
+}
+
+func (r *Registry) updateAndRegisterControlsInTopology(pluginID string, topology *report.Topology) {
+	var pluginControls []string
+	newControls := report.Controls{}
+	for controlID, control := range topology.Controls {
+		fakeID := fakeControlID(pluginID, controlID)
+		log.Debugf("plugins: replacing control %s with %s", controlID, fakeID)
+		control.ID = fakeID
+		newControls.AddControl(control)
+		pluginControls = append(pluginControls, controlID)
+		r.handlerRegistry.RegisterLocked(fakeID, r.pluginControlHandler)
```

Comment:
I think I wanted to avoid races between a goroutine that removes all the handlers of a specific plugin and the goroutine handling the control request from the plugin` like in the following scenario:

- goroutine 1:
  - plugin registry receives the report from the plugin `X`
  - plugin registry removes all the control handlers specific to the plugin `X`
  - gets preempted
- goroutine2:
  - got a control request with control ID `X~foo`
  - tries gets a handler for this control, but all the handlers were removed
  - control request fails
  - gets preempted
- goroutine 1:
  - add new handlers to the handler registry for the plugin `X` controls.

So I see two solutions here:

- Batch method for handler registry like you and I suggested in the other comment.
- Do a symmetric diff of the set of already registered handlers and the set of handlers to be registered (or do two relative complements). Then remove handlers that should be removed and add handlers that should be added.

I think I'd prefer the first solution, it's simpler.